// @walmart/wmreact-analytics/lib/backplane/collector-context.js
"use strict";

exports.__esModule = true;

var _objectWithoutProperties2 = require("babel-runtime/helpers/objectWithoutProperties");

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require("babel-runtime/helpers/possibleConstructorReturn");

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require("babel-runtime/helpers/inherits");

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
Provides a mechanism to percolate more context information down
into the subtree.
@examples
```jsx
<CollectorContext productId={123213}>
  <YourApp />
</CollectorContext>
```
@component CollectorContext
@import {CollectorContext}
*/

/* eslint prefer-const: 0 */
var CollectorContext = function (_React$Component) {
  (0, _inherits3.default)(CollectorContext, _React$Component);

  function CollectorContext(props) {
    (0, _classCallCheck3.default)(this, CollectorContext);
    return (0, _possibleConstructorReturn3.default)(this, _React$Component.call(this, props));
  }

  CollectorContext.prototype.getChildContext = function getChildContext() {
    var analytics = this.context.analytics;
    // do nothing if no analytics hooked up

    if (analytics !== undefined) {
      var _props = this.props,
          children = _props.children,
          props = (0, _objectWithoutProperties3.default)(_props, ["children"]); //eslint-disable-line no-unused-vars

      return {
        analytics: Object.assign({}, analytics, {
          context: Object.assign({}, analytics.context, props)
        })
      };
    }

    return null;
  };

  CollectorContext.prototype.render = function render() {
    return this.props.children;
  };

  return CollectorContext;
}(_react2.default.Component);

exports.default = CollectorContext;


CollectorContext.contextTypes = {
  analytics: _propTypes2.default.object
};

CollectorContext.childContextTypes = {
  analytics: _propTypes2.default.object
};

CollectorContext.propTypes = process.env.NODE_ENV !== "production" ? {
  children: _propTypes2.default.object
} : {};
