// @walmart/electrode-redux-router-engine/lib/utils/context-provider.js
"use strict";

/* NOTE: This file uses node 4 features not babel.
  For converting JSX to js code during development use:
  https://babeljs.io/repl/
*/
const React = require("react");
const PropTypes = require("prop-types");

class ContextProvider extends React.Component {
  getChildContext() {
    const {
      requestHeaders,
      requestUrl,
      requestPathParams,
      requestQryParams,
      seoTags,
      request
    } = this.props;

    return {
      // context data required by headers
      requestHeaders,
      requestUrl,
      requestPathParams,
      requestQryParams,
      // context data required by seo meta
      seoTags,
      // Electrode Platform context app object
      // - Should contain request object
      // - On client side this is not necessary exist so React code should never
      //   access this.context.app.request unless it checks this.context.app first.
      app: {
        request
      }
    };
  }

  render() {
    return this.props.children;
  }
}

ContextProvider.childContextTypes = {
  requestHeaders: PropTypes.object,
  requestUrl: PropTypes.object,
  requestPathParams: PropTypes.object,
  requestQryParams: PropTypes.object,
  seoTags: PropTypes.object,
  app: PropTypes.object
};

module.exports = ContextProvider;
