// @walmart/electrode-ui-logger/src/client.js
"use strict";
/*
 * This file is intended to be executed by browsers only
 */

/* eslint-disable consistent-return, no-magic-numbers */

const stringify = require("json-stringify-safe");
const Config = require("@walmart/electrode-ui-config");
const fetch = require("@walmart/electrode-fetch").fetch;
const Log = require("./logger");

Log._stream = [];
Log.fetch = fetch;

Log.setFetch = function(_fetch) {
  Log.fetch = _fetch;
};

Log._flush = function() {
  if (Log._stream.length > 0) {
    const payload = {
      credentials: "include",
      disableAnalytics: true,
      method: "POST",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json"
      },
      body: stringify(Log._stream)
    };

    Log._stream = [];

    return Log.fetch(Config.fullApiPath("/logger"), payload);
  }
};

Log.setLogInterval = function(_interval) {
  Log._interval = _interval;

  if (Log._poll) {
    clearInterval(Log._poll);
  }

  Log._poll = setInterval(Log._flush, Log._interval);
};

Log.setLogInterval(10000);

Log._addLogMessage = function(data) {
  Log._stream.push(data);
};

module.exports = Log;
