// @walmart/electrode-index-page/lib/hapi/default-options.js
"use strict";

function defaultOptions(env) {
  env = env || process.env;

  return {
    pageTitle: "Untitled Electrode Web Application",
    metaTags: {
      app: {
        charSet: { charset: "utf-8" },
        httpEquiv: { "http-equiv": "X-UA-Compatible", content: "IE=edge,chrome=1" },
        viewport: "width=device-width, initial-scale=1.0"
      },
      seo: {}
    },
    unbundledJS: {
      firstScripts: [],
      enterHead: [],
      preBundle: [],
      postBundle: []
    },
    unbundledStyle: null,
    typekitId: "fqp0lia",
    expData: null,
    webpackDev: env.WEBPACK_DEV === "true",
    renderJS: true,
    serverSideRendering: true,
    iconStats: "dist/server/iconstats.json",
    criticalCSS: "dist/js/critical.css",
    devServer: {
      protocol: "http",
      host: "localhost",
      port: env.WEBPACK_DEV_MIDDLEWARE === "true" ? env.PORT || "3000" : "2992",
      jsBundleName: "bundle.dev.js",
      cssBundleName: "style.css"
    },
    bundleProps: {},
    paths: {},
    cdn: {
      enabled: false,
      mapping: null, // file path of assets.json
      assetsPath: process.cwd() // directory under which to look for config/assets.json
    },
    stats: "", // optional, path of stats.json,
    statsDll: "dist/server/stats.dll.json",
    excludeTorbitContentCheck: false,
    envInfo: {
      APP_SHA: env.APP_SHA,
      APP_VERSION: env.APP_VERSION
    },
    expoPrefix: null // optional, for expo cookies prefix
  };
}

module.exports = defaultOptions;
