// @walmart/electrode-error-page/lib/error-page/generate.js
"use strict";

/* eslint-disable max-len */

var _require = require("./constants"),
    HTTP_500 = _require.HTTP_500,
    STACK_TRACES_MARK = _require.STACK_TRACES_MARK;

var renderErrorPageHtml = require("./error-page").renderErrorPageHtml;

function getBeaconCode(statusCode) {
  return "_bcq && _bcq.push([\"_tagAction\",\"ErrorPage\",\"ERRORPAGE_VIEW\",\"pgv.err.vww.pgl\",[[\"er\",{\"ht\": " + statusCode + "}]]])";
}

function getResponse(request, routeOptions, statusCode) {
  statusCode = statusCode || HTTP_500;
  var errorPage = renderErrorPageHtml(statusCode, routeOptions.excludeTorbitContentCheck);
  var sendBeacon = getBeaconCode(statusCode);
  return {
    status: statusCode,
    html: "<!DOCTYPE html><html lang=\"en\">\n<head>\n<link rel=\"dns-prefetch\" href=\"beacons.walmart.com\"/>\n<meta charSet=\"utf-8\"/>\n<meta http-equiv=\"X-UA-Compatible\" content=\"IE=edge,chrome=1\"/>\n<meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\"/>\n<meta name=\"robots\" content=\"noindex\" />\n<link rel=\"stylesheet\" id=\"bogel-font\" href=\"//i5.walmartimages.com/dfw/63fd9f59-6ace/57132652-ed71-42ee-9bc8-c256b0073184/v1/BogleWeb.css\"/>\n</head>\n<body><script>" + sendBeacon + "</script><div>" + errorPage + "</div>" + STACK_TRACES_MARK + "</body></html>"
  };
}

function responseForError(request, routeOptions, err) {
  return getResponse(request, routeOptions, err.status);
}

function responseForBadStatus(request, routeOptions, data) {
  return getResponse(request, routeOptions, data.status);
}

module.exports = {
  renderErrorPageHtml: renderErrorPageHtml,
  getBeaconCode: getBeaconCode,
  responseForError: responseForError,
  responseForBadStatus: responseForBadStatus
};
