// @walmart/wmreact-analytics/src/backplane/collector-context.jsx
/* @flow */
/* eslint prefer-const: 0 */
import React from "react";
import PropTypes from "prop-types";

/**
Provides a mechanism to percolate more context information down
into the subtree.
@examples
```jsx
<CollectorContext productId={123213}>
  <YourApp />
</CollectorContext>
```
@component CollectorContext
@import {CollectorContext}
*/
export default class CollectorContext extends React.Component {
  constructor(props: Object): void {
    super(props);
  }
  getChildContext(): {
    analytics: {
      context: Object
    }
  } {
    const { analytics } = this.context;
    // do nothing if no analytics hooked up
    if (analytics !== undefined) {
      const { children, ...props } = this.props; //eslint-disable-line no-unused-vars
      return {
        analytics: {
          ...analytics,
          context: {
            ...analytics.context,
            ...props
          }
        }
      };
    }

    return null;
  }
  render(): ReactElement {
    return this.props.children;
  }
}

CollectorContext.contextTypes = {
  analytics: PropTypes.object
};

CollectorContext.childContextTypes = {
  analytics: PropTypes.object
};

CollectorContext.propTypes = {
  children: PropTypes.object
};
