// @walmart/electrode-cookies/lib/index.js
"use strict";

const tryCatch = require("@walmart/try-catch");
const safeGet = require("lodash/get");
const reduce = require("lodash/reduce");
const getRequest = require("./get-request");

const replacers = { "(": "%28", ")": "%29" };

const encodeKey = key =>
  key.replace(/[^#$&+\^`|]/g, encodeURIComponent).replace(/[\(\)]/g, m => replacers[m]);

const cookies = {
  get: (key, options) => {
    const request = getRequest(options, "get");
    if (!request) return undefined; // w/o request there's nothing we can do.

    options = options || {};

    if (options.matchSubStr) {
      const substring = options.skipEncoding === true ? key : encodeKey(key);

      const NOT_FOUND = -1;

      return tryCatch(() =>
        reduce(
          request.state,
          (result, value, k) => {
            if (k.indexOf(substring) > NOT_FOUND) {
              result[k] =
                options.skipEncoding === true || value === undefined
                  ? value
                  : decodeURIComponent(value);
            }
            return result;
          },
          {}
        )
      );
    }

    return tryCatch(() => {
      const value = request.state[encodeKey(key)];
      return value === undefined ? undefined : decodeURIComponent(value);
    });
  },

  set: (key, value, options) => {
    const request = getRequest(options, "set");

    if (!request) return; // w/o request there's nothing we can do.

    options = options || {};

    const MSEC = 1000;

    const setOptions = {
      path: options.path || "/",
      ttl: options.expires && options.expires * MSEC,
      isHttpOnly: options.httpOnly,
      isSecure: options.secure,
      domain: options.domain,
      strictHeader: safeGet(options, "strictHeader", true)
    };

    if (!request.app.replyStates) {
      request.app.replyStates = {};
    }

    if (options.skipEncoding !== true) {
      key = encodeKey(key);
      value = (typeof value === "string" ? value : JSON.stringify(value))
        // \--9 being exclude chars - to 9, with 9 being before :
        .replace(/[^!#&-+\--9<-\[\]-~]/g, encodeURIComponent);
    }

    if (options.forceAuthEncoding) {
      value = value.replace(/[+/]/g, encodeURIComponent);
    }

    request.app.replyStates[key] = { value, options: setOptions };
  },

  expire: (key, options) => {
    cookies.set(key, "x", Object.assign({}, options, { expires: 0 }));
  }
};

module.exports = cookies;
