// @walmart/electrode-logging/lib/default-fields.js
"use strict";

const standardFields = ["request", "response", "context", "err"];

const stdRequestFields = ["method", "path", "id"];
const verboseRequestFields = stdRequestFields.concat("info", "params", "headers");

const stdResponseFields = ["statusCode"];
const verboseResponseFields = stdResponseFields.concat("settings", "headers");

const stdContextFields = ["topTxId", "txId", "parentTxId", "level", "subtype"];

const verboseContextFields = [
  "sessionId",
  "topTxId",
  "tx.status",
  "isTopLevel",
  "txId",
  "parentTxId",
  "state",
  "isRunning",
  "level",
  "duration",
  "start",
  "subtype",
  "cachedSubtype"
];

const errFields = [
  "url",
  "method",
  "headers",
  "errObj",
  "status",
  "statusText",
  "stack",
  "message",
  "name"
];

const fields = {
  standardFields: standardFields,
  request: {
    std: stdRequestFields,
    verbose: verboseRequestFields
  },
  response: {
    std: stdResponseFields,
    verbose: verboseResponseFields
  },
  context: {
    std: stdContextFields,
    verbose: verboseContextFields
  },
  err: {
    std: errFields,
    verbose: errFields
  }
};

module.exports = exports = fields;
