// @walmart/electrode-config/lib/wml-providers.js
"use strict";

const Confippet = require("electrode-confippet");

const oneOpsProfile = () => {
  if (process.env.ONEOPS_ENVPROFILE) {
    return process.env.ONEOPS_ENVPROFILE.toLowerCase();
  } else {
    return "";
  }
};

const oneOpsEnvironment = () => {
  if (process.env.ONEOPS_ENVIRONMENT) {
    return process.env.ONEOPS_ENVIRONMENT.toLowerCase();
  } else {
    return "";
  }
};

module.exports = function () {
  return {
    providers: {
      "oneopsDefault": {
        name: "oneops-default",
        order: 165,
        filter: ["{{oneOpsProfile}}"],
        type: Confippet.providerTypes.optional
      },
      "oneopsProfile": {
        name: "oneops-{{oneOpsProfile}}",
        order: 166,
        filter: ["{{oneOpsProfile}}"],
        type: Confippet.providerTypes.warn
      },
      "oneopsProfileInstance": {
        name: "oneops-{{oneOpsProfile}}-{{instance}}",
        order: 167,
        filter: ["{{oneOpsProfile}}", "{{instance}}"],
        type: Confippet.providerTypes.optional
      },
      "oneopsEnvironment": {
        name: "oneops-{{oneOpsEnvironment}}",
        order: 168,
        filter: ["{{oneOpsEnvironment}}"],
        type: Confippet.providerTypes.optional
      },
      "oneopsEnvironmentInstance": {
        name: "oneops-{{oneOpsEnvironment}}-{{instance}}",
        order: 169,
        filter: ["{{oneOpsEnvironment}}", "{{instance}}"],
        type: Confippet.providerTypes.optional
      }
    },
    context: {
      oneOpsProfile: oneOpsProfile(),
      oneOpsEnvironment: oneOpsEnvironment(),
      defaultType: Confippet.providerTypes.warn
    }
  };
};
