// @walmart/feds-error-page/lib/components/error-links.js
"use strict";

exports.__esModule = true;

var _extends2 = require("babel-runtime/helpers/extends");

var _extends3 = _interopRequireDefault(_extends2);

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _automationId = require("../utils/automation-id");

var _statusCodes = require("../enums/status-codes");

var _statusCodes2 = _interopRequireDefault(_statusCodes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _renderReloadLink = function _renderReloadLink(_ref) {
  var onClick = _ref.onClick,
      dataAutomationId = _ref.dataAutomationId;

  var Link = onClick ? "span" : "a";
  var linkProps = onClick ? { onClick: onClick } : { href: "" };

  return _react2.default.createElement(
    Link,
    (0, _extends3.default)({
      id: "retry-link",
      className: "error-page__link"
    }, linkProps, (0, _automationId.getDataAutomationIdPair)("reload", dataAutomationId)),
    "Try again"
  );
};

var _renderSeparator = function _renderSeparator(is500) {
  return is500 ? " or go back to " : "Go to ";
};

var ErrorLinks = function ErrorLinks(props) {
  var statusCode = props.statusCode,
      suppressHomeLink = props.suppressHomeLink,
      dataAutomationId = props.dataAutomationId;

  var IS_500 = statusCode === _statusCodes2.default["500"];

  return _react2.default.createElement(
    "div",
    { className: "error-page__links" },
    IS_500 && _renderReloadLink(props),
    !suppressHomeLink && _renderSeparator(IS_500),
    !suppressHomeLink && _react2.default.createElement(
      "a",
      (0, _extends3.default)({
        className: "error-page__link",
        href: "/"
      }, (0, _automationId.getDataAutomationIdPair)("home", dataAutomationId)),
      "Walmart.com"
    ),
    !suppressHomeLink && "."
  );
};

ErrorLinks.displayName = "ErrorLinks";

ErrorLinks.propTypes = process.env.NODE_ENV !== "production" ? {
  dataAutomationId: _propTypes2.default.string,
  onClick: _propTypes2.default.func,
  statusCode: _propTypes2.default.number.isRequired,
  suppressHomeLink: _propTypes2.default.bool
} : {};

ErrorLinks.defaultProps = {
  dataAutomationId: "error-ErrorPage-link",
  suppressHomeLink: false
};

exports.default = ErrorLinks;
