// @walmart/feds-error-page/lib/components/error-page.js
"use strict";

exports.__esModule = true;

var _extends2 = require("babel-runtime/helpers/extends");

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require("babel-runtime/helpers/objectWithoutProperties");

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _spark = require("./spark");

var _spark2 = _interopRequireDefault(_spark);

var _errorLinks = require("./error-links");

var _errorLinks2 = _interopRequireDefault(_errorLinks);

var _statusMessages = require("../enums/status-messages");

var _statusMessages2 = _interopRequireDefault(_statusMessages);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ErrorPage = function ErrorPage(props) {
  var dataAutomationId = props.dataAutomationId,
      onClick = props.onClick,
      statusCode = props.statusCode,
      suppressHeader = props.suppressHeader,
      suppressHomeLink = props.suppressHomeLink,
      otherProps = (0, _objectWithoutProperties3.default)(props, ["dataAutomationId", "onClick", "statusCode", "suppressHeader", "suppressHomeLink"]);


  return _react2.default.createElement(
    "div",
    null,
    !suppressHeader && _react2.default.createElement(
      "header",
      { className: "error-page__header" },
      _react2.default.createElement(
        "a",
        { href: "/" },
        _react2.default.createElement(_spark2.default, { title: "Walmart. Save Money. Live Better." })
      )
    ),
    _react2.default.createElement(
      "div",
      (0, _extends3.default)({ className: "error-page__content" }, otherProps),
      _react2.default.createElement(_spark2.default, { size: 64, className: "error-page__logo" }),
      _react2.default.createElement(
        "h1",
        { className: "error-page__title" },
        "Sorry..."
      ),
      _react2.default.createElement(
        "p",
        { className: "error-page__message" },
        _statusMessages2.default[statusCode]
      ),
      _react2.default.createElement(_errorLinks2.default, {
        dataAutomationId: dataAutomationId,
        onClick: onClick,
        statusCode: statusCode,
        suppressHomeLink: suppressHomeLink
      })
    )
  );
};

ErrorPage.displayName = "ErrorPage";

ErrorPage.propTypes = process.env.NODE_ENV !== "production" ? {
  dataAutomationId: _propTypes2.default.string,
  onClick: _propTypes2.default.func,
  statusCode: _propTypes2.default.number.isRequired,
  suppressHeader: _propTypes2.default.bool,
  suppressHomeLink: _propTypes2.default.bool
} : {};

ErrorPage.defaultProps = {
  dataAutomationId: "error-ErrorPage-link",
  suppressHeader: false,
  suppressHomeLink: false
};

exports.default = ErrorPage;
